flycs_sdk/functions.py

```diff
@@ -0,0 +1,166 @@
+        description: Optional[str] = None,
+        static: Optional[bool] = True,
+        destination_data_mart: Optional[str] = None,
+    ):
+        """Create a Function object.
+
+        :param name: name of the function
+        :type name: str
+        :param query: SQL body of the function
+        :type query: str
+        :param version: version of the function
+        :type version: str
+        :param argument_list: the list of arguments of the function
+        :type argument_list: List[Argument]
+        :param return_type: the SQL return type of the function
+        :type return_type: str
+        :param language: the language of the function, defaults to sql
+        :type language: Optional[str]
+        :param description: description of the function, defaults to None
+        :type description: Optional[str], optional
+        """
+        super().__init__(
+            name=name,
+            query=query,
+            version=version,
+            static=static,
+            destination_data_mart=destination_data_mart,
+        )
+        self.destination_table = None
+        self.dependencies = []
+        self.parsing_dependencies = []
+        self.argument_list = argument_list
+        self.description = description
+        self.return_type = return_type
+        self.language = language
+
+    @classmethod
+    def from_dict(cls, d: dict):
+        """Create a Function object from a dictionary created with the to_dict method.
+
+        :param d: source dictionary
+        :type d: dict
+        :return: Function
+        :rtype: Function
+        """
+        function = cls(
+            name=d.get("NAME", ""),
+            query=d["QUERY"],
+            version=d["VERSION"],
+            description=d.get("DESCRIPTION"),
+            static=d.get("STATIC", True),
+            # encrypt=d.get("ENCRYPT", False),
+            destination_data_mart=d.get("DESTINATION_DATA_MART"),
+            argument_list=[Argument.from_dict(a) for a in d.get("ARGUMENT_LIST") or []],
+            return_type=d.get("RETURN_TYPE"),
+            language=d.get("LANGUAGE", "sql"),
+        )
+        function.destination_table = d.get("DESTINATION_TABLE")
+        function.dependencies = [
+            Dependency.from_dict(x) for x in d.get("DEPENDS_ON") or []
+        ]
+        function.parsing_dependencies = [
+            Dependency.from_dict(x) for x in d.get("PARSING_DEPENDS_ON") or []
+        ]
+        return function
+
+    def to_dict(self) -> dict:
+        """
+        Serialize the Function to a dictionary object.
+
+        :return: the Function as a dictionary object.
+        :rtype: Dict
+        """
+        return {
+            "NAME": self.name,
+            "QUERY": self.query,
+            "VERSION": self.version,
+            "DESCRIPTION": self.description,
+            "DESTINATION_TABLE": self.destination_table,
+            "KIND": self.kind,
+            "ENCRYPT": self.encrypt,
```

Comment:
Is it normal that we export ENCRYPT, but don't use it in the class?

Reply:
Yeah it's because it is a member of the base class, so it is set to None by default. For functions this attribute does not make much sense, so maybe we should add another level in the class hierarchy to split this from the base class.

Reply:
was thinking the same. Seems we can split the base class in two different more narrow interface. The QueryBase class is really for anything that contains a query.

Reply:
I had another look at it. Encrypt is actually already an optional field, so I would say it's fine to just skip it and do not use it in the `to_dict` and `from_dict` methods.